Create the Concept of Multiple EventPipe Sessions

jorive opened the pull request:
This is the initial work to enable https://github.com/dotnet/coreclr/issues/15377

## What's here?

- A lot of code move/split. Some important moves:
  - `EventPipe` has a colection of `EventPipeSessions` instead of a single session.
  - `EventPipeSession` now owns a `EventPipeBufferManager` and a `EventPipeFile`
  - `EventPipeThread` now owns a collection of { EventPipeBufferManager, EventPipeBuffer }, and a collection of { EventPipeBufferManager, EventPipeBufferList }
  - There is a cap on the max number of `EventPipeSession` (64 sessions)
  - `EventPipeProvider` and `EventPipeEvent` use a 64-bit mask to keep track of the sessions that are listening to provider/events.

## What's pending?

- There is an assertion (bug?) in `EventPipeThread` when disabling a second concurrent session that needs investigation:
```log
Assert failure(PID 22748 [0x000058dc], Thread: 25352 [0x6308]): m_pWriteBuffers->GetCount() == 1

CORECLR! EventPipeThread::~EventPipeThread + 0x3E (0x00007ff9`b015632e)
CORECLR! EventPipeThread::`scalar deleting destructor' + 0x17 (0x00007ff9`b01574d7)
CORECLR! EventPipeThread::Release + 0x4A (0x00007ff9`b0162f2a)
CORECLR! ReleaseEventPipeThreadRef + 0x13 (0x00007ff9`b0162f53)
CORECLR! FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef>::DoRelease + 0x16 (0x00007ff9`b015c646)
CORECLR! BaseHolder<EventPipeThread *,FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef>,0,&CompareDefault<EventPipeThread *> >::Release + 0x4A (0x00007ff9`b0162eca)
CORECLR! BaseHolder<EventPipeThread *,FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef>,0,&CompareDefault<EventPipeThread *> >::~BaseHolder<EventPipeThread *,FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEv + 0.x1C (0x00007ff9`b0155dbc)
CORECLR! BaseWrapper<EventPipeThread *,FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef>,0,&CompareDefault<EventPipeThread *> >::~BaseWrapper<EventPipeThread *,FunctionBase<EventPipeThread *,&AcquireEventPipeThreadRef,&Release + 0x1C (0x00007ff9`b0155dec)
CORECLR! Wrapper<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef,0,&CompareDefault<EventPipeThread *>,1>::~Wrapper<EventPipeThread *,&AcquireEventPipeThreadRef,&ReleaseEventPipeThreadRef,0,&CompareDefault<EventPipeThread *>,1> + 0x1C (0x00007ff9`b015611c)
CORECLR! EventPipeBufferList::~EventPipeBufferList + 0x23 (0x00007ff9`b0156153)
    File: .\src\vm\eventpipebuffermanager.cpp Line: 51
    Image: .\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe
```
- Review a potential race condition between the EventPipe::WriteEvent and EventPipe::Disable. Maybe we need to add a `SimpleRWLock` to the collection of sessions in `EventPipe`.
- More testing
- Review all multi-threading aspects and contracts/assertions
- Currently, we are using a hash map to keep track of sessions on EventPipe, but maybe we could change it to an array of pointers to `EventPipeSession` objects.
- When a session encounters an error on EventPipeSession::ThreadProc, we should notify EventPipe to tear down the session from its list and do the appropriate cleanup.
- Optimizations, mainly around duplicated sessions/providers, filtering events, etc.

noahfalk commented on src/vm/eventpipesession.cpp:
```diff
@@ -89,4 +86,157 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+// }
+
+const EventPipeProviderConfiguration RundownProviders[] = {
+    {W("Microsoft-Windows-DotNETRuntime"), 0x80020138, static_cast<unsigned int>(EventPipeEventLevel::Verbose), NULL},       // Public provider.
```
We should document what this magic number is and pull it out into a define. It looks like it is:
LoaderKeyword(0x8) | JitKeyword(0x10) | NgenKeyword (0x20) | unused_keyword (0x100) | JittedMethodILToNativeMapKeyword (0x20_000) | ThreadTransferKeyword (0x8000_0000)

noahfalk commented on src/vm/eventpipesession.cpp:
```diff
@@ -89,4 +86,157 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+                    Config.GetKeywords(),
+                    (EventPipeEventLevel)Config.GetLevel(),
+                    Config.GetFilterData());
+                m_pProviderList->AddSessionProvider(pProvider);
```
Above I worked out what these rundown keywords were and it certainly looks like a bunch of stuff that logs normally in addition to rundown specific things. That would make a little more sense why the old code created a thread-specific filter. We should probably re-instate that (@cshung may have done that already?). If for whatever reason don't wind up doing that then we need to do something to deal with the multi-threaded calls to EventPipeFile::WriteEvent that are about to start happening after RundownEnabled = true.

noahfalk commented on src/vm/eventpipesession.cpp:
```diff
@@ -89,4 +86,157 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+        Thread *pEventThread,
+        StackContents *pStack)
+{
+    return m_pBufferManager->WriteEvent(pThread, *this, event, payload, pActivityId, pRelatedActivityId);
```
I assume there is some TODO work in order to filter events to specific sessions that enabled them. If that was already occuring somewhere I didn't catch it : )

jorive replied:
You are right, there is an implicit TODO here (I added it to the PR description but forgot to add it here).

noahfalk commented:
In terms of the upcoming Crst work I'd suggest not adding a new Crst if possible and instead use the EventPipe::m_lock mostly as is. I think this is doable but let me know if this doesn't pan out. There are two major categories of operations the lock is used for that I am aware of:
a) 'fast work' - These are all configuration changes such as add session, remove session, add provider, add event, enable event, disable event. None of these things should require calling back into managed code under the lock nor waiting for any other thread to finish work. These operations all presumably complete in nanoseconds or microseconds and they occur infrequently enough that I hope contention is rare, even across all sessions combined.
b) 'slow work' - Disabling a session needs to wait for threads to finish writing, queues to drain, and rundown to happen. This could take seconds (or even minutes) and if two sessions tried to do it at the same time it would be pretty easy for one of them to block the other substantially. 

My suggestion is that we stop holding the EventPipe::m_lock during the slow work so that it is reasonable to keep using it for everything else. For Disable() I think this is how we could get similar invariants to what we had before without holding the lock the whole time:
- Make sure threads can't call Enable and Disable at the same time. 
  - No Thread can call Disable during Enable because Enable doesn't provide back the session id until it is done and the caller needs that to call Disable.
- Make sure that multiple threads can't try to run the disable code concurrently
  - Add a state field to the session that progresses from enabled -> disable_in_progress -> disabled
  - Use EventPipe::m_lock to update that flag at the beginning and end of disable, but don't hold it in the middle.
  - If a second thread tries to call Disable() and the state is already disable_in_progress then exit EventPipe::m_lock and wait on a ManualResetEvent allocated for the session. At the same time the first Disable() thread updates the state to disabled it can set the ManualResetEvent allowing any additional waiting threads to proceed.

jorive commented on src/vm/eventpipeconfiguration.cpp:
```diff
@@ -149,14 +140,14 @@ void EventPipeConfiguration::DeleteProvider(EventPipeProvider *pProvider)
 }
 
-bool EventPipeConfiguration::RegisterProvider(EventPipeProvider &provider, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
+bool EventPipeConfiguration::RegisterProvider(EventPipeSession *pSession, EventPipeProvider &provider, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
```
Good call :)

cshung commented on src/vm/eventpipe.cpp:
```diff
@@ -245,90 +256,70 @@ EventPipeSessionID EventPipe::Enable(
+        // FIXME: What's the story here? SampleProfiler should be enabled only once.
+        SampleProfiler::SetSamplingRate(static_cast<unsigned long>(profilerSamplingRateInNanoseconds));
+
+        // TODO: Maybe "rundownEnabled" param should be user input.
```
Thanks for thinking about it - I think we want to make sure it is a user input.
https://github.com/dotnet/diagnostics/issues/189

jorive replied:
Maybe we also want to control the sampling rate (currently hardcoded).

noahfalk replied:
It will be challenging to let the users control the sampling rate because if two sessions both want to control it who wins? I think we should find out what AI profiler sets it to now and assuming that is a reasonable default we should hardcode that number. Its not a great story, but I think it is an OK one for this release.

jorive replied:
I'm thinking we can leave this for when we have the protocol ready. I will open an issue later.

mjsabby commented:
I'd love to see this work to allow the Profiler API getting EventPipe events and also other tools that may want to use EventPipe to write to multiple files for scenarios where you'd like symbols data in a single file throughout the application lifecycle and period data dumps for other events.

jorive commented on src/vm/eventpipe.cpp:
```diff
@@ -382,74 +398,21 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
+        s_pConfig->Disable(pEventPipeProviderCallbackDataQueue);
 
-        // Delete the flush timer.
+        //  TODO: Delete IFF there are not more sessions.
```
Easiest right now would be one IPC streaming thread per session. We could optimize later :)

cshung commented on src/vm/eventpipe.cpp:
```diff
@@ -193,36 +190,49 @@ void EventPipe::Shutdown()
-    // Flush process write buffers to make sure other threads can see the change.
-    s_pConfig = NULL;
-    s_pBufferManager = NULL;
     FlushProcessWriteBuffers();
```
FlushProcessWriteBuffer() should happen right after setting the pointers to NULL?

cshung commented on src/vm/eventpipe.cpp:
```diff
@@ -193,36 +190,49 @@ void EventPipe::Shutdown()
+    delete pSessions;
+
     delete s_pEventSource;
     s_pEventSource = NULL;
```
Why s_pEventSource work differently from s_pConfig and s_pSessions?

jorive replied:
Agree... currently this call what it does is to subscribe the "Microsoft-DotNETCore-EventPipe" provider to the session.

cshung commented on src/vm/eventpipe.cpp:
```diff
@@ -343,34 +336,33 @@ void EventPipe::Disable(EventPipeSessionID id)
+    if (s_pConfig != nullptr)
     {
         // Disable the profiler.
         SampleProfiler::Disable();
```
We should not disable the SampleProfiler when only one session disable?

jorive replied:
Yes, this is incomplete. I'll push a commit with @noahfalk suggestion.

cshung commented on src/vm/fastserializer.cpp:
```diff
@@ -17,8 +17,8 @@ IpcStreamWriter::IpcStreamWriter(uint64_t id, IpcStream *pStream) : _pStream(pSt
         NOTHROW;
-        GC_TRIGGERS;
-        MODE_PREEMPTIVE;
+        GC_NOTRIGGER;
```
This change is concerning, can you explain why this change is necessary?

jorive replied:
Agreed, and we should review our contracts.
This is the assertion I was getting without these changes:
```log
Assert failure(PID 16760 [0x00004178], Thread: 22220 [0x56cc]): CONTRACT VIOLATION by FastSerializer::WriteTag at "c:\coreclr\src\vm\fastserializer.cpp" @ 250

MODE_PREEMPTIVE encountered while thread is in cooperative state.

                        CONTRACT in FastSerializer::WriteTag at "c:\coreclr\src\vm\fastserializer.cpp" @ 250
                        CONTRACT in FastSerializer::WriteObject at "c:\coreclr\src\vm\fastserializer.cpp" @ 163
                        CONTRACT in EventPipeFile::Flush at "c:\coreclr\src\vm\eventpipefile.cpp" @ 117
                        CONTRACT in EventPipeBufferManager::WriteAllBuffersToFile at "c:\coreclr\src\vm\eventpipebuffermanager.cpp" @ 515
                        CONTRACT in EventPipeSession::WriteAllBuffersToFile at "c:\coreclr\src\vm\eventpipesession.cpp" @ 269
                        CONTRACT in ThreadProc at "c:\coreclr\src\vm\eventpipesession.cpp" @ 129

We can't find the violated contract. Look for an old-style non-holder-based contract.


CORECLR! CONTRACT_ASSERT + 0x345 (0x00007ffe`86f9a265)
CORECLR! EEContract::DoChecks + 0x2E6 (0x00007ffe`870ef826)
CORECLR! FastSerializer::WriteTag + 0x11B (0x00007ffe`87a7f5eb)
CORECLR! FastSerializer::WriteObject + 0x230 (0x00007ffe`87a7efc0)
CORECLR! EventPipeFile::Flush + 0x14B (0x00007ffe`87a8329b)
CORECLR! EventPipeBufferManager::WriteAllBuffersToFile + 0x609 (0x00007ffe`87a99539)
CORECLR! EventPipeSession::WriteAllBuffersToFile + 0x262 (0x00007ffe`878d0982)
CORECLR! <lambda_6ce8943d6e0712d6a69eb273347d9526>::operator() + 0x5D (0x00007ffe`878ceefd)
CORECLR! EventPipe::RunWithCallbackPostponed<<lambda_6ce8943d6e0712d6a69eb273347d9526> > + 0x46 (0x00007ffe`878ce3e6)
CORECLR! ThreadProc + 0x34C (0x00007ffe`878d06ac)
    File: c:\coreclr\src\vm\fastserializer.cpp Line: 250
    Image: C:\coreclr\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe
```

cshung commented on src/vm/eventpipesession.cpp:
```diff
@@ -44,23 +84,151 @@ EventPipeSession::~EventPipeSession()
+            {
+                {
+                    // Switch to pre-emptive mode so that this thread doesn't starve the GC
+                    GCX_PREEMP();
```
Is this just copy-and-paste from SampleProfiler?

I am not sure why are we doing it - the contract indicates this thread is MODE_PREEMPTIVE, not sure why we need to do an explicit switch here.

cshung commented on src/vm/eventpipesession.cpp:
```diff
@@ -86,7 +255,162 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+        // Reset the event before shutdown.
+        m_threadShutdownEvent.Reset();
+
+        // The sampling thread will watch this value and exit
```
Typo: The command should mention the IPC streaming thread.

cshung commented on src/vm/eventpipesession.cpp:
```diff
@@ -81,12 +248,167 @@ EventPipeSessionProvider* EventPipeSession::GetSessionProvider(EventPipeProvider
+        m_ipcStreamingEnabled = false;
+
+        // Wait for the sampling thread to clean itself up.
+        m_threadShutdownEvent.Wait(INFINITE, FALSE /* bAlertable */);
```
This is where you could have a deadlock.
Suppose one thread reach here and is waiting for the streaming thread to complete, and the streaming thread cannot complete because it cannot acquire the EventPipe::GetLock(), both thread will not be able to proceed and you have a deadlock.

jorive replied:
Yup, that's what I'm seeing.

noahfalk commented on src/vm/eventpipe.cpp:
```diff
@@ -343,34 +335,34 @@ void EventPipe::Disable(EventPipeSessionID id)
         GC_TRIGGERS;
         MODE_ANY;
-        PRECONDITION(GetLock()->OwnedByCurrentThread());
+        PRECONDITION(IsLockOwnedByCurrentThread());
```
I'm fine if you want to do it in a later PR, but something should probably track breaking the lock when doing Disable(), otherwise Disable() on session A will be able to block all other sessions from doing work for a long time.

jorive replied:
I'll postpone it to future PR.

noahfalk commented on src/vm/eventpipe.cpp:
```diff
@@ -392,216 +384,34 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
-        pSession = s_pSession;
-    }
-    return pSession;
+    if (s_pSessions == nullptr)
```
Lets add an assert that EventPipe::m_lock is being held to this function.

noahfalk commented on src/vm/eventpipe.cpp:
```diff
@@ -732,66 +533,68 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
     }
     CONTRACTL_END;
 
+    if (s_pConfig == nullptr || s_pSessions == nullptr)
```
@cshung - Not a high pri issue, but one spot that might be interesting to poke for reliability is synchronization between writer threads and shutdown. Any potential issues here pre-dated @jorive 's change but the fact that Brian originally put similar checks in the code suggests that some thread could try to call WriteEvent() after EventPipe::Shutdown() had occured. And if the call can come afterwards, there is a decent chance it could also happen concurrently. Even if s_pConfig isn't nullptr at this instruction, maybe it becomes null 20 lines lower in the function and I'm guessing the results wouldn't be good : )

noahfalk commented on src/vm/eventpipe.cpp:
```diff
@@ -732,66 +533,68 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
+        if (pSession == nullptr)
+            continue;
+
+        if (!pSession->RundownEnabled())
```
A performance issue to track if it isn't already on the list - I suspect that one session triggering rundown causes those events to spill over into other sessions that happen to have overlapping keywords enabled. Its probably a low pri issue.

jorive commented:
@noahfalk @cshung I have updated the PR with the "last" set of suggestions. There is more work to be done here, including testing. Could you please review my changes one more time, and let me know if I can squash/merge this PR? This PR is already big enough that I would like to get it merge and start fixing bugs in different PRs.
In addition, I rebased, and updated the PR description above.
Cheers,
-José

jorive commented:
@noahfalk @cshung I would like to review the debug-contracts later on with both of you.

noahfalk commented on src/vm/sampleprofiler.cpp:
```diff
@@ -103,34 +120,35 @@ void SampleProfiler::Disable()
         // Synchronization of multiple callers occurs in EventPipe::Disable.
-        PRECONDITION(EventPipe::GetLock()->OwnedByCurrentThread());
+        PRECONDITION(EventPipe::IsLockOwnedByCurrentThread());
+        PRECONDITION(s_RefCount >= 0);
```
I don't think this precondition is true? If you create one session that does not use the sampling profiler and then disable it DisableInternal() will call here and s_RefCount will be 0.

noahfalk replied:
Nevermind, I misread it as s_RefCount > 0. What you wrote looks correct. Ignore silly ole me : )

jorive replied:
`PRECONDITION(s_RefCount >= 0);` Zero is acceptable :)

noahfalk commented on src/vm/sampleprofiler.cpp:
```diff
@@ -103,34 +120,35 @@ void SampleProfiler::Disable()
-    // when profiling is disabled.
-    s_profilingEnabled = false;
+        // Reset the event before shutdown.
+        s_threadShutdownEvent.Reset();
```
I would reset the event prior to creating to the sampling thread. Otherwise any error that causes the thread to exit prematurely will set the event, then it will be reset here, and the Wait below will block forever.

noahfalk commented on src/vm/eventpipeconfiguration.cpp:
```diff
@@ -482,6 +427,8 @@ EventPipeEventInstance *EventPipeConfiguration::BuildEventMetadataEvent(EventPip
     }
     CONTRACTL_END;
 
+    // FIXME: Per-session tracking whether event metadata has been emitted.
```
Maybe I misunderstood, I thought based on the meeting today that this was done now?

noahfalk commented on src/vm/eventpipe.cpp:
```diff
@@ -193,19 +213,27 @@ void EventPipe::Shutdown()
         GC_TRIGGERS;
         MODE_ANY;
+        // These 3 pointers are initialized once on EventPipe::Initialize
+        // FIXME: Why is shutdown called more than once?
```
There are 17 FIXMEs and 15 TODOs in the comments of this PR. Before checking in it would be good to triage these and either remove + fix them, remove + file an issue to fix them later, or decide that they can be removed with no further action taken.

jorive replied:
Resolving this as I opened https://github.com/dotnet/coreclr/issues/24753 to track all FIXME/TODO.

jorive commented on src/vm/eventpipeevent.cpp:
```diff
@@ -111,74 +99,19 @@ BYTE *EventPipeEvent::BuildMinimumMetadata()
+void EventPipeEvent::RefreshState(
+    uint64_t sessionId,
+    INT64 sessionKeywords,
+    EventPipeEventLevel sessionLevel)
```
Left over. I'll remove it.

jorive commented on src/vm/eventpipeevent.cpp:
```diff
@@ -111,74 +99,19 @@ BYTE *EventPipeEvent::BuildMinimumMetadata()
     return minmumMetadata;
 }
 
-EventPipeProvider* EventPipeEvent::GetProvider() const
```
I disagree with the "imaginary style point" for trivial getters (just return a field), but this PR is big enough that I will undo it.

noahfalk replied:
No qualms if you have different style preferences than me, I wouldn't have said anything if the code was new. What raised the bar high enough that I commented is that someone else already placed the code in the .cpp and defering to the existing precedent, even when you don't personally prefer the style, is the only thing that prevents unending code churn.

jorive commented on src/vm/eventpipeconfiguration.cpp:
```diff
@@ -281,96 +281,78 @@ EventPipeProvider *EventPipeConfiguration::GetProviderNoLock(const SString &prov
 }
 
-void EventPipeConfiguration::Enable(EventPipeSession *pSession, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
+void EventPipeConfiguration::Enable(EventPipeSession &session, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
```
To minimize checking for null to bail out earlier and avoid AV, and adding asserts around them.

jorive commented on src/vm/eventpipeprovider.cpp:
```diff
@@ -131,29 +103,55 @@ bool EventPipeProvider::EventEnabled(INT64 keywords, EventPipeEventLevel eventLe
-    m_enabled = providerEnabled;
-    m_keywords = keywords;
-    m_providerLevel = providerLevel;
+    m_sessions ^= sessionId;
```
Yes, it is a left over when I was trying to use the function to toggle configuration on/off.

jorive commented on src/vm/eventpipeprovider.cpp:
```diff
@@ -131,29 +103,55 @@ bool EventPipeProvider::EventEnabled(INT64 keywords, EventPipeEventLevel eventLe
 
-    RefreshAllEvents();
+    // Set the provider level to "Log Always" or the biggest verbosity.
+    m_providerLevel = (providerLevel == EventPipeEventLevel::LogAlways) ? providerLevel : ((providerLevel < m_providerLevel) ? m_providerLevel : providerLevel);
```
From https://github.com/dotnet/coreclr/blob/master/src/vm/eventpipeprovider.cpp#L125
```
The event is enabled if:

  - The provider is enabled.
  - The event keywords are unspecified in the manifest (== 0) or when masked with the enabled config are != 0.
  - The event level is LogAlways or the provider's verbosity level is set to greater than the event's verbosity level in the manifest.
```

jorive replied:
When multiple sessions are enabling the same provider with potentially different keywords and levels, I want to enable the union of all, so all sessions can get their respective data.

jorive replied:
Do you have a better idea?

noahfalk replied:
I think you can just eliminate the condition for LogAlways. LogAlways is the lowest level so it is guaranteed that any provider will have a log level that is >= LogAlways. 

m_providerLevel = (providerLevel < m_providerLevel) ? m_providerLevel : providerLevel);

jorive commented on src/vm/eventpipeprovider.cpp:
```diff
@@ -131,29 +103,55 @@ bool EventPipeProvider::EventEnabled(INT64 keywords, EventPipeEventLevel eventLe
+
+    if (m_sessions & sessionId)
+        m_sessions &= ~sessionId;
+    return PrepareCallbackData(nullptr);
```
Yes, currently, I just remove the session from the bitmask. Ideally, I should keep the union of the keywords/level per provider for all remaining active sessions.

For example, two sessions below enable the same provider with different keywords and level

    Session A: Enable("Microsoft-Windows-DotNETRuntime:0x0000000000008001:4")
    Session B: Enable("Microsoft-Windows-DotNETRuntime:0x000000000000FFFF:5")

When disabling session A, I cannot just reset all keywords without considering the active sessions.

jorive replied:
I can open an issue and handle it in a different PR, or would you prefer here?

noahfalk replied:
Opening an issue is fine by me.

cshung commented on src/vm/eventpipeconfiguration.cpp:
```diff
@@ -256,8 +256,8 @@ EventPipeProvider *EventPipeConfiguration::GetProviderNoLock(const SString &prov
-        GC_TRIGGERS;
-        MODE_PREEMPTIVE;
+        GC_NOTRIGGER;
+        MODE_ANY;
```
As you are relaxing the contract, it would be nice to comment why we need to relax them?

jorive replied:
CI uncovered different path that triggered assertions that I did not want to handle them in this PR, therefore I reverted some of the contract changes.

jorive replied:
This is one of the assertions:
```
Assert failure(PID 8432 [0x000020f0], Thread: 8436 [0x20f4]): CONTRACT VIOLATION by EventPipeConfiguration::GetProvider at \"f:\\workspace\\_work\\1\\s\\src\\vm\\eventpipeconfiguration.cpp\" @ 245

GC_TRIGGERS encountered in a GC_NOTRIGGER scope

                        CONTRACT in EventPipeConfiguration::GetProvider at \"f:\\workspace\\_work\\1\\s\\src\\vm\\eventpipeconfiguration.cpp\" @ 245
VIOLATED-->  CONTRACT in EventPipe::GetProvider at \"f:\\workspace\\_work\\1\\s\\src\\vm\\eventpipe.cpp\" @ 524
                        CONTRACT in EventPipeInternal::GetProvider at \"f:\\workspace\\_work\\1\\s\\src\\vm\\eventpipeinternal.cpp\" @ 129
                        CONTRACT in DispatchCallSimple at \"f:\\workspace\\_work\\1\\s\\src\\vm\\callhelpers.cpp\" @ 185
                        CONTRACT in MethodTable::RunClassInitEx at \"f:\\workspace\\_work\\1\\s\\src\\vm\\methodtable.cpp\" @ 3331
                        GCX_COOP in MethodTable::DoRunClassInitThrowing at \"f:\\workspace\\_work\\1\\s\\src\\vm\\methodtable.cpp\" @ 3438
                        CONTRACT in MethodTable::DoRunClassInitThrowing at \"f:\\workspace\\_work\\1\\s\\src\\vm\\methodtable.cpp\" @ 3436
                        CONTRACT in MethodTable::CheckRunClassInitThrowing at \"f:\\workspace\\_work\\1\\s\\src\\vm\\methodtable.cpp\" @ 3735
                        CONTRACT in DynamicHelperFixup at \"f:\\workspace\\_work\\1\\s\\src\\vm\\prestub.cpp\" @ 2846
                        GCX_PREEMP_THREAD_EXISTS in DynamicHelperWorker at \"f:\\workspace\\_work\\1\\s\\src\\vm\\prestub.cpp\" @ 3209
                        CONTRACT in MethodDescCallSite::CallTargetWorker at \"f:\\workspace\\_work\\1\\s\\src\\vm\\callhelpers.cpp\" @ 344
                        GCX_COOP in Assembly::ExecuteMainMethod at \"f:\\workspace\\_work\\1\\s\\src\\vm\\assembly.cpp\" @ 1656
                        CONTRACT in Assembly::ExecuteMainMethod at \"f:\\workspace\\_work\\1\\s\\src\\vm\\assembly.cpp\" @ 1640
                        GCX_COOP in CorHost2::ExecuteAssembly at \"f:\\workspace\\_work\\1\\s\\src\\vm\\corhost.cpp\" @ 436
                        CONTRACT in CorHost2::ExecuteAssembly at \"f:\\workspace\\_work\\1\\s\\src\\vm\\corhost.cpp\" @ 385



CORECLR! <no symbol> + 0x0 (0x00007ff8`3a584646)
CORECLR! GetCLRRuntimeHost + 0x4F8B6 (0x00007ff8`3a65bca6)
CORECLR! CreateCLRProfiling + 0x24024 (0x00007ff8`3ad11a84)
CORECLR! GetCLRRuntimeHost + 0x36FE6E (0x00007ff8`3a97c25e)
CORECLR! CreateCLRProfiling + 0x694D0 (0x00007ff8`3ad56f30)
<no module>! <no symbol> + 0x0 (0x00007ff7`db46cf40)
<no module>! <no symbol> + 0x0 (0x00000098`0e97bc80)
<no module>! <no symbol> + 0x0 (0x00000098`0e97b9e0)
<no module>! <no symbol> + 0x0 (0x0000024c`09ee2c50)
<no module>! <no symbol> + 0x0 (0x00007ff7`db8f32f8)
    File: f:\\workspace\\_work\\1\\s\\src\\vm\\eventpipeconfiguration.cpp Line: 245
    Image: C:\\dotnetbuild\\work\\7abaa591-329d-4487-9783-ef582e2757a5\\Payload\\CoreRun.exe


Return code:      1
Raw output file:      C:\\dotnetbuild\\work\\7abaa591-329d-4487-9783-ef582e2757a5\\Work\\4db27c6d-4903-4693-b97f-a9e0ad004cc1\\Exec\\tracing\\eventcounter\\Reports\\tracing.eventcounter\\incrementingpollingcounter\\incrementingpollingcounter.output.txt
Raw output:
BEGIN EXECUTION
 \"C:\\dotnetbuild\\work\\7abaa591-329d-4487-9783-ef582e2757a5\\Payload\\corerun.exe\" incrementingpollingcounter.exe 
Test Passed
Expected: 100
Actual: -1073740286
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\\dotnetbuild\\work\\7abaa591-329d-4487-9783-ef582e2757a5\\Payload
> C:\\dotnetbuild\\work\\7abaa591-329d-4487-9783-ef582e2757a5\\Work\\4db27c6d-4903-4693-b97f-a9e0ad004cc1\\Exec\\tracing\\eventcounter\\incrementingpollingcounter\\incrementingpollingcounter.cmd
Expected: True
Actual:   False
```

jorive commented on src/vm/eventpipeevent.h:
```diff
@@ -46,44 +46,99 @@ class EventPipeEvent
+        _ASSERT(m_pProvider != nullptr);
+        const bool IsProviderEnabled = m_pProvider->IsEnabled(sessionId);
+        const bool IsEventEnabled = m_pProvider->EventEnabled(m_keywords, m_level);
+        return (IsProviderEnabled && IsEventEnabled);
```
This is a TODO for subsequent PR.

jorive replied:
Added follow up task on issue below.

jorive commented on src/vm/eventpipe.cpp:
```diff
@@ -357,265 +377,102 @@ void EventPipe::Disable(EventPipeSessionID id)
-    {
-        _ASSERTE(s_fileSwitchTimerHandle == NULL);
+    EventPipeSession *pSession = nullptr;
+    if (!s_pSessions->Lookup(id, &pSession) || (pSession == nullptr))
```
Next PR

jorive commented on src/vm/eventpipeevent.h:
```diff
@@ -46,44 +46,99 @@ class EventPipeEvent
+        LIMITED_METHOD_CONTRACT;
+        _ASSERT(m_pProvider != nullptr);
+        const bool IsProviderEnabled = m_pProvider->IsEnabled(sessionId);
+        const bool IsEventEnabled = m_pProvider->EventEnabled(m_keywords, m_level);
```
Added a follow up task on issue below.

jorive commented on src/vm/eventpipeprovider.cpp:
```diff
@@ -131,29 +103,55 @@ bool EventPipeProvider::EventEnabled(INT64 keywords, EventPipeEventLevel eventLe
     //  - The event level is LogAlways or the provider's verbosity level is set to greater than the event's verbosity level in the manifest.
     return (EventEnabled(keywords) &&
-        ((eventLevel == EventPipeEventLevel::LogAlways) || (m_providerLevel >= eventLevel)));
+            ((eventLevel == EventPipeEventLevel::LogAlways) || (m_providerLevel >= eventLevel)));
```
Added comment to issue below.

sywhang commented on src/vm/eventpipe.cpp:
```diff
@@ -166,6 +164,24 @@ BYTE *EventPipeEventPayload::GetFlatData()
+void EventPipeProviderCallbackDataQueue::Enqueue(EventPipeProviderCallbackData *pEventPipeProviderCallbackData)
+{
+    SListElem<EventPipeProviderCallbackData> *listnode = new SListElem<EventPipeProviderCallbackData>(); // throws
+    listnode->m_Value = *pEventPipeProviderCallbackData;
```
when the call above fails this might AV?

sywhang replied:
nvm. that doesn't have (nothrow).

jorive commented:
@dotnet/dnceng Could you please tell me what does [this](https://dev.azure.com/dnceng/public/_build/results?buildId=202398&view=logs&jobId=59d6606d-5812-580a-cd03-8a1d307fc6d7&taskId=29c59d8b-55c0-5250-d201-03b7cfeca283&lineStart=1&lineEnd=12&colStart=1&colEnd=1) mean:
```log
##[section]Starting: Save product build as pipeline artifact
==============================================================================
Task         : Publish Pipeline Artifact
Description  : Publish Pipeline Artifact
Version      : 0.139.0
Author       : Microsoft Corporation
Help         : Publish a local directory or file as a named artifact for the current pipeline.
==============================================================================
Uploading pipeline artifact from F:\workspace\_work\1\s\bin\Product\Windows_NT.x64.Checked for build #202398
##[error]TF400677: Host b55de4ed-4b5a-4215-a8e4-0a0a5f71e7d8 does not exist.
##[section]Finishing: Save product build as pipeline **artifact**
```

riarenas commented:
@jorive That looks like https://github.com/dotnet/coreclr/issues/24787 which should be fixed as of https://github.com/dotnet/coreclr/pull/24788

MattGal commented:
@riarenas I think those issues are the workaround for the actual problem which doesn't seem to have been addressed.  I'm chatting with some VSTS folks now.

jorive commented:
@MattGal Would you know how I can re-trigger the legs, so they actually pass this step? I've tried quite few times, and I even re-based 2 days ago.

jorive commented:
Current failure was on `Build Windows_NT arm Checked`
```log
##[section]Starting: Build packages
==============================================================================
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613735)
==============================================================================
Generating script.
Script contents:
build-packages.cmd -BuildArch=arm -BuildType=Checked -OfficialBuildId=20190528.745 -ci
========================== Starting Command Output ===========================
##[command]"C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "F:\workspace\_work\_temp\06a60b1d-d8c7-4d0a-9bbf-172899491f9b.cmd""
Processing F:\workspace\_work\1\s\global.json
File 'C:\Users\vsagent\.netcoreeng\native\temp\cmake-3.11.1-win64-x64.zip' already exists, skipping download
File 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' already exists, skipping download
VERBOSE: Extracting 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' to 
'F:\workspace\_work\1\s\.tools\bin\WinShimmer'
File 'C:\Users\vsagent\.netcoreeng\native\temp\python-2.7.15-win64-x64.zip' already exists, skipping download
Native tools are available from F:\workspace\_work\1\s\.tools\bin
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=F:\workspace\_work\1\s\artifacts\toolset\1.0.0-beta.19273.11.txt /t:__WriteToolsetLocation /warnaserror F:\workspace\_work\1\s\artifacts\toolset\restore.proj
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog /bl:F:\workspace\_work\1\s\bin\Logs\build-packages.binlog /clp:Summary /logger:F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\netcoreapp2.1\Microsoft.DotNet.Arcade.Sdk.dll /nr:False /nodeReuse:false /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:Configuration=Debug /p:RepoRoot=F:\workspace\_work\1\s /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /p:__BuildOS=Windows_NT /p:ArcadeBuild=true /p:PortableBuild=true /p:FilterToOSGroup=Windows_NT /p:__BuildArch=arm /p:__BuildType=Checked /p:OfficialBuildId=20190528.745 /p:Projects=F:\workspace\_work\1\s\src\.nuget\packages.builds /warnaserror F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Build.proj
  Restore completed in 2.03 sec for F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Tools.proj.
  Restore completed in 40.32 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Jit\Microsoft.NETCore.Jit.builds.
  Restore completed in 40.84 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILDAsm\Microsoft.NETCore.ILDAsm.builds.
  Restore completed in 20.08 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Native\Microsoft.NETCore.Native.builds.
  Restore completed in 50.62 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILAsm\Microsoft.NETCore.ILAsm.builds.
  Restore completed in 5.41 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Runtime.CoreCLR\Microsoft.NETCore.Runtime.CoreCLR.builds.
  Restore completed in 5.97 ms for F:\workspace\_work\1\s\src\.nuget\packages.builds.
  Restore completed in 6.33 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.TestHost\Microsoft.NETCore.TestHost.builds.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745

Build FAILED.

F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:06.02
Build failed.
See log: F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog
ERROR: An error occurred while building packages. See log for more details:
    F:\workspace\_work\1\s\bin\Logs\build-packages.binlog
##[error]Cmd.exe exited with code '1'.
##[section]Finishing: Build packages
```

jorive commented:
Previous failure on `Build Windows_NT arm Checked`
```log
##[section]Starting: Build packages
==============================================================================
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613735)
==============================================================================
Generating script.
Script contents:
build-packages.cmd -BuildArch=arm -BuildType=Checked -OfficialBuildId=20190528.745 -ci
========================== Starting Command Output ===========================
##[command]"C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "F:\workspace\_work\_temp\06a60b1d-d8c7-4d0a-9bbf-172899491f9b.cmd""
Processing F:\workspace\_work\1\s\global.json
File 'C:\Users\vsagent\.netcoreeng\native\temp\cmake-3.11.1-win64-x64.zip' already exists, skipping download
File 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' already exists, skipping download
VERBOSE: Extracting 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' to 
'F:\workspace\_work\1\s\.tools\bin\WinShimmer'
File 'C:\Users\vsagent\.netcoreeng\native\temp\python-2.7.15-win64-x64.zip' already exists, skipping download
Native tools are available from F:\workspace\_work\1\s\.tools\bin
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=F:\workspace\_work\1\s\artifacts\toolset\1.0.0-beta.19273.11.txt /t:__WriteToolsetLocation /warnaserror F:\workspace\_work\1\s\artifacts\toolset\restore.proj
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog /bl:F:\workspace\_work\1\s\bin\Logs\build-packages.binlog /clp:Summary /logger:F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\netcoreapp2.1\Microsoft.DotNet.Arcade.Sdk.dll /nr:False /nodeReuse:false /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:Configuration=Debug /p:RepoRoot=F:\workspace\_work\1\s /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /p:__BuildOS=Windows_NT /p:ArcadeBuild=true /p:PortableBuild=true /p:FilterToOSGroup=Windows_NT /p:__BuildArch=arm /p:__BuildType=Checked /p:OfficialBuildId=20190528.745 /p:Projects=F:\workspace\_work\1\s\src\.nuget\packages.builds /warnaserror F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Build.proj
  Restore completed in 2.03 sec for F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Tools.proj.
  Restore completed in 40.32 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Jit\Microsoft.NETCore.Jit.builds.
  Restore completed in 40.84 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILDAsm\Microsoft.NETCore.ILDAsm.builds.
  Restore completed in 20.08 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Native\Microsoft.NETCore.Native.builds.
  Restore completed in 50.62 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILAsm\Microsoft.NETCore.ILAsm.builds.
  Restore completed in 5.41 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Runtime.CoreCLR\Microsoft.NETCore.Runtime.CoreCLR.builds.
  Restore completed in 5.97 ms for F:\workspace\_work\1\s\src\.nuget\packages.builds.
  Restore completed in 6.33 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.TestHost\Microsoft.NETCore.TestHost.builds.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745

Build FAILED.

F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:06.02
Build failed.
See log: F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog
ERROR: An error occurred while building packages. See log for more details:
    F:\workspace\_work\1\s\bin\Logs\build-packages.binlog
##[error]Cmd.exe exited with code '1'.
##[section]Finishing: Build packages
```

jorive commented:
@noahfalk @tommcdon I would like to merge this PR as the current failures are unrelated to my changes (they seem infrastructure failures).

tommcdon commented:
> @noahfalk @tommcdon I would like to merge this PR as the current failures are unrelated to my changes (they seem infrastructure failures).

@jorive sounds reasonable to me

jorive commented:
After one last rerun, the failure on `Build Windows_NT arm Checked: Build Packages`:
```log
##[section]Starting: Build packages
==============================================================================
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613735)
==============================================================================
Generating script.
Script contents:
build-packages.cmd -BuildArch=arm -BuildType=Checked -OfficialBuildId=20190528.745 -ci
========================== Starting Command Output ===========================
##[command]"C:\Windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "F:\workspace\_work\_temp\d8808a1f-4f5e-4cf5-872c-33d5bd7a92a8.cmd""
Processing F:\workspace\_work\1\s\global.json
File 'C:\Users\vsagent\.netcoreeng\native\temp\cmake-3.11.1-win64-x64.zip' already exists, skipping download
File 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' already exists, skipping download
VERBOSE: Extracting 'C:\Users\vsagent\.netcoreeng\native\temp\WinShimmer.zip' to 
'F:\workspace\_work\1\s\.tools\bin\WinShimmer'
File 'C:\Users\vsagent\.netcoreeng\native\temp\python-2.7.15-win64-x64.zip' already exists, skipping download
Native tools are available from F:\workspace\_work\1\s\.tools\bin
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=F:\workspace\_work\1\s\artifacts\toolset\1.0.0-beta.19273.11.txt /t:__WriteToolsetLocation /warnaserror F:\workspace\_work\1\s\artifacts\toolset\restore.proj
F:\workspace\_work\1\s\.dotnet\sdk\3.0.100-preview6-011681\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog /bl:F:\workspace\_work\1\s\bin\Logs\build-packages.binlog /clp:Summary /logger:F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\netcoreapp2.1\Microsoft.DotNet.Arcade.Sdk.dll /nr:False /nodeReuse:false /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:Configuration=Debug /p:RepoRoot=F:\workspace\_work\1\s /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /p:__BuildOS=Windows_NT /p:ArcadeBuild=true /p:PortableBuild=true /p:FilterToOSGroup=Windows_NT /p:__BuildArch=arm /p:__BuildType=Checked /p:OfficialBuildId=20190528.745 /p:Projects=F:\workspace\_work\1\s\src\.nuget\packages.builds /warnaserror F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Build.proj
  Restore completed in 2.88 sec for F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Tools.proj.
  Restore completed in 66.51 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILAsm\Microsoft.NETCore.ILAsm.builds.
  Restore completed in 20.88 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Native\Microsoft.NETCore.Native.builds.
  Restore completed in 66.51 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Jit\Microsoft.NETCore.Jit.builds.
  Restore completed in 66.5 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.ILDAsm\Microsoft.NETCore.ILDAsm.builds.
  Restore completed in 4.18 ms for F:\workspace\_work\1\s\src\.nuget\packages.builds.
  Restore completed in 4.52 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.Runtime.CoreCLR\Microsoft.NETCore.Runtime.CoreCLR.builds.
  Restore completed in 4.41 ms for F:\workspace\_work\1\s\src\.nuget\Microsoft.NETCore.TestHost\Microsoft.NETCore.TestHost.builds.
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
##[error].packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745

Build FAILED.

F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\1.0.0-beta.19273.11\tools\Version.targets(17,5): error : Invalid format of BuildNumber: 20190528.745 [F:\workspace\_work\1\s\src\.nuget\packages.builds]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:06.73
Build failed.
See log: F:\workspace\_work\1\s\artifacts\log\Debug\Build.binlog
ERROR: An error occurred while building packages. See log for more details:
    F:\workspace\_work\1\s\bin\Logs\build-packages.binlog
##[error]Cmd.exe exited with code '1'.
##[section]Finishing: Build packages
```
